base/osutils.jl

```diff
@@ -2,59 +2,95 @@
 
 const OS_NAME = ccall(:jl_get_OS_NAME, Any, ())
 
-function is_unix(os::Symbol)
-    if (os==:Windows) return false;
-    elseif (os==:Linux) return true;
-    elseif (os==:FreeBSD) return true;
-    elseif (os==:Darwin) return true;
-    else throw(ArgumentError("unknown operating system, $(repr(os))"))
-    end
-end
+"""
+    is_unix([os])
 
-function _os_test(qm,ex,test)
-    @assert qm == :?
-    @assert isa(ex,Expr)
-    @assert ex.head == :(:)
-    @assert length(ex.args) == 2
-    if test
-        return esc(ex.args[1])
+Predicate for testing if the OS provides a Unix-like interface.
+See documentation :ref:`Handling Operating System Variation <man-handling-operating-system-variation>`\ .
+"""
+function is_unix(os::Symbol = OS_NAME)
+    if is_windows(os)
+        return false
+    elseif is_linux(os) || is_bsd(os)
+        return true
     else
-        return esc(ex.args[2])
+        throw(ArgumentError("unknown operating system \"$os\""))
     end
 end
 
-macro windows(qm,ex)
-    _os_test(qm, ex, OS_NAME===:Windows)
-end
-macro unix(qm,ex)
-    _os_test(qm, ex, is_unix(OS_NAME))
-end
-macro osx(qm,ex)
-    _os_test(qm, ex, OS_NAME===:Darwin)
-end
-macro linux(qm,ex)
-    _os_test(qm, ex, OS_NAME===:Linux)
-end
+"""
+    is_linux([os])
 
-macro windows_only(ex)
-    @windows? esc(ex) : nothing
-end
-macro unix_only(ex)
-    @unix? esc(ex) : nothing
-end
-macro osx_only(ex)
-    @osx? esc(ex) : nothing
-end
-macro linux_only(ex)
-    @linux? esc(ex) : nothing
+Predicate for testing if the OS is a derivative of Linux.
+See documentation :ref:`Handling Operating System Variation <man-handling-operating-system-variation>`\ .
+"""
+is_linux(os::Symbol = OS_NAME) = (os == :Linux)
+
+"""
+    is_bsd([os])
+
+Predicate for testing if the OS is a derivative of BSD.
+See documentation :ref:`Handling Operating System Variation <man-handling-operating-system-variation>`\ .
+"""
+is_bsd(os::Symbol = OS_NAME) = (os == :FreeBSD || os == :OpenBSD || os == :NetBSD || os == :Darwin)
+
+"""
+    is_windows([os])
+
+Predicate for testing if the OS is a derivative of Microsoft Windows NT.
+See documentation :ref:`Handling Operating System Variation <man-handling-operating-system-variation>`\ .
+"""
+is_windows(os::Symbol = OS_NAME) = (os == :Windows)
+
+"""
+    is_apple([os])
+
+Predicate for testing if the OS is a derivative of Apple Macintosh OS X or Darwin.
```

Comment:
> Do we want this to return true for iOS?

I think yes. I believe `iOS` is a very similar kernel and user-space to the desktop platform.